### ISSUE TITLE
connectMenu Silently Swallows Event (But it Would be Really Handy if it Passed it Along)

### ISSUE THREAD

**Issue by machineghost** [vkbansal/react-contextmenu#191]
## Version
1.9.1

## Test Case

    connectMenu(`foo)(MenuToConnect);

## Steps to reproduce

Connect a menu, trigger the menu, try to use the event that triggered the menu to decide what should be in the ContextMenu

## Expected Behavior

The (dynamic, ie. connected via `connectMenu`) ContextMenu has access to the event that triggered the showing of it.

## Actual Behavior

All that gets passed is:

     <Child {...this.props} id={menuId} trigger={this.state.trigger} />;

Meanwhile `e` gets lost back in `handleShow` (in https://github.com/vkbansal/react-contextmenu/blob/2ef4aae1aa624275f1c9f50d67417d39b2788d22/src/connectMenu.js).

It would be really nice if instead `handleShow` kept the event in state, then passed it to the `Child` when it rendered:

            handleShow = (e) => {
                // ...
                this.setState({ e, trigger: filteredData }); // NOTE the added "e, "
            }

            handleHide = () => {
                this.setState({ trigger: null });
            }

            render() {
                return <Child {...this.props} e={this.state.e} id={menuId} trigger={this.state.trigger} />;
            }
        };

This way if I have fifty buttons on the page, instead of having to wrap each of them with a separate `<ContextMenuTrigger>`, I can just have one `<ContextMenuTrigger>` wrapping the lot of them, and then I can look at `e.target.id` (or `e.whatever`) to decide what options the resulting menu should offer.

Also this would eliminate many of the needs for nested triggers: instead of having to nest triggers one would only need the outermost trigger, which could then check the event to see if an inner (nested) element was responsible.

I would be happy to submit a PR adding this functionality if desired.

**Comment by vkbansal** [comment]
connectMenu is experimental.

@CarstenWickner Can you help out?

**Comment by CarstenWickner** [comment]
Hi @machineghost,
the whole point of the `filteredData` part was to not expose more than is necessary.
Exposing the whole event would mean to provide references to heaps of unnecessary data.
However, I can appreciate the idea of forwarding the reference from e.target to reduce the need for too many nested triggers.

I'd propose to only include this additional reference to the `target` and nothing more, or what other field from the event do you think would be meaningful?

Happy to review a pull request if you create one.

**Comment by machineghost** [comment]
I just don't see the point of hiding basic browser/language features.  Like if you've got 50 private variables that the user shouldn't be able to mess with, by all means don't expose them.  But what do you gain from hiding the browser event object?  How does hiding parts of it make the library better?

**Comment by CarstenWickner** [comment]
First of all it's not "my library", just the guy who implemented the "connectMenu" PR (based on someone else's suggestion event) -- so I'm only expressing my personal opinion.

I'm with you on the topic of not needing to hide standard events. That being said, have you tried your suggested code snippet? Because the event reaching the `handleShow(e)` method is not actually a standard one but is being created in in the [src/action.js](https://github.com/vkbansal/react-contextmenu/blob/8985f3bd14afb82ed04aa2edd1ec60f2dae66ad5/src/actions.js#L15).
```javascript
event = new window.CustomEvent(eventName, { detail: opts });
```

Those `opts` being passed in contain some information collected in the [src/ContextMenuTrigger](https://github.com/vkbansal/react-contextmenu/blob/8985f3bd14afb82ed04aa2edd1ec60f2dae66ad5/src/ContextMenuTrigger.js#L102)
```
let data = callIfExists(this.props.collect, this.props);
let showMenuConfig = {
    position: { x, y },
    target: this.elem,
    id: this.props.id,
    data
};
...
showMenu(showMenuConfig);
```
I.e. the only additional information on there that might be useful is the clicked `position`. Not sure whether forwarding the specified `target` would help you here if you have only a single trigger containing everything.

Long story short, your proposed one-line change wouldn't do you much good here, we'd have to specifically forward any information you're interested in from the actual event (which could be one of multiple possibilities).

Therefore I asked what specific information you're looking for here. ;)

**Comment by CarstenWickner** [comment]
Hi @machineghost,
I created a [proposal](https://github.com/vkbansal/react-contextmenu/compare/master...CarstenWickner:issue%23191) for exposing the target component (without anything else from the event for now).

But there should definitely be a meaningful example and possibly test for this.
I.e. I'm still open to review a pull request. 😉

**Comment by machineghost** [comment]
Philosophically I still see no reason why you wouldn't just do:

    var originalEvent = theMouseDownEventThatTriggeredThings;

    // then pass originalEvent into dispatchGlobalEvent and do:
    event = new window.CustomEvent(eventName, { detail: opts, ... originalEvent });

    // or maybe (since I'm not sure the above will work)
    event = Object.assign(new window.CustomEvent(eventName, { detail: opts }), originalEvent);

but practically speaking providing the `target` alone would certainly handle my use case.

Thanks!

**Comment by wootencl** [comment]
After reading through this issue + discussion a few times I think there may have been a bit of a disconnect on what @machineghost was requesting and what @CarstenWickner had in mind. Please correct me if I'm wrong though. 

I _believe_ the request was around wanting access to the event that triggered the context menu opening (the one seen [here](https://github.com/vkbansal/react-contextmenu/blob/master/src/ContextMenuTrigger.js#L95)). Admittedly though I would agree that access to the entirety of this event seems a little unnecessary and may involve `e.persist()` if the user wanted to use it down the road. I would think a happy medium would be to pass the `target` of that event. This would allow for the event delegation I _think_ @machineghost is referring to in his initial post. Current thoughts would be to just extract the `target` in the above segment of code and include it along side the `data` extracted by the user specified `collect` (e.g. - `{ ...data, target: event.target }`). This could then be accessed in the `ContextMenu`'s `render` method via `this.props.trigger.target`.

As this would be a fairly small change I'd be happy to make a PR if that's of interest.

**Comment by machineghost** [comment]
Philosophically I object to my tools hiding information from me that I would otherwise have if I wasn't using the tool.  It feels ... patronizing.

Practically, the use case that spawned this ticket could have been solved with just the `target`, so that would certainly be an improvement.
